### ISSUE TITLE
Acquire exchange currency deposit address

### ISSUE THREAD

**Issue by PhinixPhire** [ccxt-dev/ccxt#288]
Is there a way for me to pull the deposit address of a particular currency from a ccxt exchange object? 

The purpose is to be able to move funding between exchanges automatically in preparation for an event. The exchange.withdraw(currency, amount, address) works great for pulling money out, but I've been maintaining a database of deposit addresses manually entered that I'd love to populate programmatically instead. 

Thanks!

**Comment by kroitor** [comment]
CCXT supports all methods for all exchanges, but all exchanges are not standardized at all, in fact they are all different, therefore we have to unify some methods into a common subset of names that will be consistent across all exchanges. So, in others words, having all their public and private methods, we already support generating or retrieving a deposit address with any exchange, if the exchange in question offers that itself. But not all of them are unified yet (it's a work in progress). We will be implementing the common unified `deposit()` method soon. 

In the meantime you can use their custom private API methods which are all implemented in CCXT, for obtaining a deposit address. A quick example of how you would do it for an arbitrary exchange (just to make it clear):

```Python
import ccxt
bitfinex = ccxt.bitfinex ({
    "apiKey": "YOUR_API_KEY",
    "secret": "YOUR_SECRET",
})
response = bitfinex.private_post_deposit_new({
    'method': 'bitcoin',  # their unique currency naming, like "bitcoin", "ethereum"...
    'wallet_name': 'exchange',
    'renew': 0,  # a value of 1 will generate a new address
})
print(response)  # this will print a raw unparsed response from the exchange
```

Actually, while writing the example above I realized that I just implemented a first deposit method for Bitfinex, so I included it as a reference implementation in CCXT, and you can now do it easily with the unified method as shown below:

```Python
deposit = bitfinex.deposit('BTC')  # the unified method works with unified currency names
print(deposit)  # will print a nice parsed reply, with an 'address' field in it
# or
print(bitfinex.deposit('BTC', {'renew': 1}))  # will generate a new address
```

**Comment by kroitor** [comment]
(You will need to wait for about 5 minutes for the build of next version to arrive). Cheers!

**Comment by kroitor** [comment]
Ok, the build is there, please update to v1.9.65+, all of the above should work. I'm closing this for now until we have further updates on deposit unification. Thx!

**Comment by PhinixPhire** [comment]
Related question, sorry to revive the old thread. I wanted to inquire further about other non-unified methods and how to implement them. 

In this particular case, I need to fetch the HitBTC "Main" account balance. 
Their documentation on fetching "Main" balance can be found here: https://hitbtc.com/api#paymentbalance
The documentation for fetching "Trading" balance is here: https://hitbtc.com/api#tradingbalance

But both appear to be named the same (balance) with a different parent (trading vs payment... )
Request: GET /api/1/trading/balance
Request: GET /api/1/payment/balance

In ccxt's code, fetch_balance appears to point to trading...privateGet**Trading**Balance(), but I don't know where that code resides to investigate further:
```
    def fetch_balance(self, params={}):
        self.load_markets()
        balances = self.privateGetTradingBalance()
        result = {'info': balances}
        for b in range(0, len(balances)):
            balance = balances[b]
            code = balance['currency']
            currency = self.common_currency_code(code)
            account = {
                'free': float(balance['available']),
                'used': float(balance['reserved']),
                'total': 0.0,
            }
            account['total'] = self.sum(account['free'], account['used'])
            result[currency] = account
        return self.parse_balance(result)
```

Basically, how would I go about pursuing the "Main"(/Payment) balance as can be retrieved by the above linked documentation? 

Sorry to bug you again. I tried to research this myself as best I could. I don't fully understand the non-unified methods and am having a hard time grasping the concepts.

**Comment by kroitor** [comment]
Hi! )

No problem, your question is fine. But it was asked and answered previously )

https://github.com/ccxt-dev/ccxt/issues/237#issuecomment-332354313
https://github.com/ccxt-dev/ccxt/wiki/Manual#api-method-naming-conventions

Basically, the method you are looking for does not exist upon launch, and it gets created by the exchange upon instantiation, like so:

```
GET /api/1/trading/balance gets converted to a callable method: 
        trading_get_balance ()

GET /api/1/payment/balance gets converted to a callable method: 
        payment_get_balance ()
```

↑ From the above examples the naming convention for generated methods should be clear. Each URL from api schema of each exchange gets its own callable method with a unique name.

**Comment by kroitor** [comment]
To get a list of all available methods with an exchange instance, you can simply do the following:

```
console.log (new ccxt.kraken ())   // JavaScript
print (dir (ccxt.hitbtc ()))        # Python
var_dump (new \ccxt\okcoinusd ()); // PHP
```

**Comment by kroitor** [comment]
I just added a parameterized version of it for you, so you can now fetch any balance like so:

```
hitbtc.fetch_balance()  # returns trading account balance by default
hitbtc.fetch_balance({'type': 'trading'})  # returns trading account balance
hitbtc.fetch_balance({'type': 'payment'})  # returns main account balance
```

Please update to v.1.9.96+ in order to use that (will be available in 5 minutes).

**Comment by kroitor** [comment]
The new build is there, and it now allows you to fetch either of you balances ) We will be happy if you report back, whether it helps you or not ) Thx!

**Comment by PhinixPhire** [comment]
@kroitor - I'm not sure if I'm doing something wrong. I'm trying to accomplish this:

```
print hitbtc.payment_post_transfer_to_trading ({
    'amount': hitbtc.fetch_balance({'type': 'payment'}),
    'currency_code': 'AEON',
})
```

But this part breaks:

`print hitbtc.fetch_balance({'type': 'payment'})`

With the following Traceback:

```
Traceback (most recent call last):
  File "C:\Users\Kenneth\eclipse-workspace\Chank\CCXTtests.py", line 91, in <module>
    print hitbtc.fetch_balance({'type': 'payment'})
  File "C:\Python27\lib\site-packages\ccxt\exchanges.py", line 11683, in fetch_balance
    'free': float(balance['cash']),
KeyError: 'cash'
```

**Comment by PhinixPhire** [comment]
I should mention, I upgraded to 1.9.100: 
```
      Successfully uninstalled ccxt-1.9.65
Successfully installed ccxt-1.9.100
```

**Comment by kroitor** [comment]
@PhinixPhire i must have left something over... I'll take a look and get back to you.

**Comment by kroitor** [comment]
Well, I think I nailed it at last ) Sorry for the inconvenience. Please update once more (1.9.101) and retry. It must work now! %) Thx!

**Comment by PhinixPhire** [comment]
You nailed it! 
The following code is working great:

```
print hitbtc.payment_post_transfer_to_trading ({
    'amount': hitbtc.fetch_balance({'type': 'payment'}).get('AEON')['free'],
    'currency_code': 'AEON',
})
```

Thanks so much. Great work!

**Comment by Bomper** [comment]
Note that to [get the main account balance](https://github.com/ccxt/ccxt/issues/288#issuecomment-335549106) on HitBTC, your API key needs to have the "Payment information" permission, not just the "Order book, History, **Trading balance**" permission. Blame [the Chinese](https://www.reddit.com/r/Monero/comments/7jn790/how_do_you_send_monero_to_binance/).

**Comment by kroitor** [comment]
@Bomper thanks for the helpful hint, very true.

**Comment by asocanari** [comment]
> I just added a parameterized version of it for you, so you can now fetch balance for either of your HitBTC accounts like so:
> 
> ```
> hitbtc.fetch_balance()  # returns trading account balance by default
> hitbtc.fetch_balance({'type': 'trading'})  # returns trading account balance
> hitbtc.fetch_balance({'type': 'payment'})  # returns main account balance
> ```
> 
> Please update to v.1.9.96+ in order to use that (will be available in 5 minutes).

HI there, is this functionality maybe broken ? I am trying to do 

`
        let ex = await exchange.getExchangeObject('hitbtc2');
        balances = await ex.fetchBalance({'type': 'trading'});
`
which works fine so far, but 

`
        let ex = await exchange.getExchangeObject('hitbtc2');
        balances = await ex.fetchBalance({'type': 'payment'});
`
returns 

(node:28915) UnhandledPromiseRejectionWarning: TypeError: this[method] is not a function
    at hitbtc2.fetchBalance (/srv/XXX/node_modules/ccxt/js/hitbtc2.js:666:43)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:189:7)
(node:28915) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:28915) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

On version ccxt: '1.17.461'
